src/lib/transition_frontier_controller_tests/test_catchup_scheduler.ml

```diff
@@ -0,0 +1,400 @@
+      in
+      let _catchup_breadcrumbs_reader, catchup_breadcrumbs_writer =
+        Strict_pipe.create Synchronous
+      in
+      Thread_safe.block_on_async_exn (fun () ->
+          let open Deferred.Let_syntax in
+          let%bind frontier = setup_random_frontier () in
+          let scheduler =
+            Catchup_scheduler.create ~logger ~frontier ~time_controller
+              ~catchup_job_writer ~catchup_breadcrumbs_writer
+          in
+          let randomly_chosen_breadcrumb =
+            Transition_frontier.all_breadcrumbs frontier
+            |> List.permute |> List.hd_exn
+          in
+          let%bind upcoming_breadcrumbs =
+            Deferred.all
+            @@ Quickcheck.random_value ~size:2
+                 (Quickcheck_lib.gen_imperative_list
+                    ( randomly_chosen_breadcrumb |> return
+                    |> Quickcheck.Generator.return )
+                    (gen_breadcrumb ~logger ~accounts_with_secret_keys))
+          in
+          let dangling_breadcrumb = List.nth_exn upcoming_breadcrumbs 1 in
+          let dangling_transition =
+            Transition_frontier.Breadcrumb.transition_with_hash
+              dangling_breadcrumb
+          in
+          Catchup_scheduler.watch scheduler ~timeout_duration
+            ~transition:dangling_transition ;
+          let result_ivar = Ivar.create () in
+          Strict_pipe.Reader.iter_without_pushback catchup_job_reader
+            ~f:(fun catchup_transition ->
+              Ivar.fill result_ivar catchup_transition )
+          |> don't_wait_for ;
+          let%map catchup_transition = Ivar.read result_ivar in
+          assert (
+            External_transition.Verified.equal
+              (With_hash.data dangling_transition)
+              (With_hash.data catchup_transition) ) ;
+          Strict_pipe.Writer.close catchup_breadcrumbs_writer ;
+          Strict_pipe.Writer.close catchup_job_writer )
+
+    let%test_unit "if a linear sequence of transitions come in order and the \
+                   missing node is received before the timeout expires, the \
+                   timeout would be canceled" =
+      let _catchup_job_reader, catchup_job_writer =
+        Strict_pipe.create Synchronous
+      in
+      let catchup_breadcrumbs_reader, catchup_breadcrumbs_writer =
+        Strict_pipe.create Synchronous
+      in
+      Thread_safe.block_on_async_exn (fun () ->
+          let open Deferred.Let_syntax in
+          let%bind frontier = setup_random_frontier () in
+          let scheduler =
+            Catchup_scheduler.create ~logger ~frontier ~time_controller
+              ~catchup_job_writer ~catchup_breadcrumbs_writer
+          in
+          let randomly_chosen_breadcrumb =
+            Transition_frontier.all_breadcrumbs frontier
+            |> List.permute |> List.hd_exn
+          in
+          let%bind upcoming_breadcrumbs =
+            Deferred.all
+            @@ Quickcheck.random_value ~size:4
```

Comment:
The code for generating breadcrumbs is quite repetitive. Can you make a convenience function that creates a linear list of breadcrumbs using an arbitrary breadcrumb, accounts_with_secret_keys and logger.

Do this for generating a k tree of breadcrumbs.

Also, make these changes for other functions generating breadcrumbs in this library.